### PR TITLE
Enhanced extension dependencies validation

### PR DESCRIPTION
**Pull request by aloubyansky**
As a follow up to https://github.com/quarkusio/quarkus/pull/14463
Here is an enhanced version of the extension dependencies validator.
In addition to making sure for every dependency on the runtime extension artifact there is the corresponding dependency on the deployment extension artifact, this version also checks that:
* extension deployment artifacts do not appear among the runtime dependencies;
* for every dependency on a deployment extension artifact there is the corresponding dependency on the runtime extension artifact;
* a deployment artifact does not depend on an extension's runtime artifact w/o it being present on the runtime classpath.

FYI @stuartwdouglas @geoand

@geoand and @iocanel please review the kubernetes client related adjustments.

### CODE REVIEW

**Comment by geoand** [comment]
Thanks @aloubyansky !

@iocanel can you do the review of the client? My back log of stuff has grown significantly today and I don't want to hold up this until I can get to it

File: test-framework/kubernetes-client/pom.xml [hunk 1]
```diff
@@ -21,10 +21,12 @@
             <groupId>io.fabric8</groupId>
             <artifactId>kubernetes-server-mock</artifactId>
             <exclusions>
+            <!--
                 <exclusion>
                     <groupId>com.squareup.okhttp3</groupId>
                     <artifactId>okhttp</artifactId>
                 </exclusion>
+                -->
```
Comment (gsmet) [inline test-framework/kubernetes-client/pom.xml:29]: Either we drop it the exclusion entirely or at least we add a comment explaining the why. Because this is puzzling.

**Review by gsmet — CHANGES_REQUESTED** [review]
The changes look good to me but... we would need this PR to be organized so that the build is not broken (typically we could have a commit fixing all the issues BEFORE the commit implementing the checks).

Maybe it's already working this way but the commits look a bit suspicious :).

**Comment by aloubyansky** [comment]
A PR adjusting the deps before the enhanced validation is doable. I'm still waiting for the feedback on the kubernetes client deps.

**Comment by geoand** [comment]
@iocanel can you give this a look please?

**Comment by geoand** [comment]
@iocanel any chance you can take a look at this?

File: extensions/kubernetes-client/runtime-internal/pom.xml [hunk 2]
```diff
@@ -13,11 +13,33 @@
...
     <dependencies>
         <dependency>
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-core</artifactId>
+            <artifactId>quarkus-arc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.fabric8</groupId>
+            <artifactId>kubernetes-client</artifactId>
+            <exclusions>
```
Comment (iocanel) [inline extensions/kubernetes-client/runtime-internal/pom.xml:25]: The exclusions below should no longer be needed as of kubernetes-client 5.0.1 (currently using 5.0.2).

**Review by iocanel — APPROVED** [review]
The Kubernetes related bits look good to me.
I added an optional remark about some exclusions we could possibly remove, but given that this might be a little out of scope, we could definitely omit it for now.

**Comment by aloubyansky** [comment]
Thanks @iocanel I'll rebase and organize it in two PRs as suggested by @gsmet

File: extensions/jdbc/jdbc-h2/runtime/pom.xml [hunk 3]
```diff
@@ -13,6 +13,14 @@
     <name>Quarkus - JDBC - H2 - Runtime</name>
     <description>Connect to the H2 database via JDBC</description>
     <dependencies>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-arc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-datasource</artifactId>
```
Comment (aloubyansky) [inline extensions/jdbc/jdbc-h2/runtime/pom.xml:22]: @stuartwdouglas I haven't yet looked deeper but just as a quick fix to make the enhanced extension dependency validation to pass I had to add `quarkus-datasource` extension as a runtime dependency to most of the jdbc extensions. Otherwise, it would complain that the deployment artifact depends on `quarkus-datasource-deployment` through `quarkus-devservices-XXX` but the `quarkus-datasource` isn't found among the runtime dependencies.

**Comment by aloubyansky** [comment]
@gsmet the commits are re-ordered as you requested. But I still want to clarify the `quarkus-datasource` dependencies in the jdbc extensions.

**Comment by aloubyansky** [comment]
@gsmet This one should be good now.

File: test-framework/kubernetes-client/pom.xml [hunk 4]
```diff
@@ -21,10 +21,6 @@
             <groupId>io.fabric8</groupId>
             <artifactId>kubernetes-server-mock</artifactId>
             <exclusions>
-                <exclusion>
-                    <groupId>com.squareup.okhttp3</groupId>
-                    <artifactId>okhttp</artifactId>
-                </exclusion>
```
Comment (gsmet) [inline test-framework/kubernetes-client/pom.xml:27]: Why did you remove this one?

Comment (aloubyansky) [reply test-framework/kubernetes-client/pom.xml:27]: As a consequence of refactoring of the kubernetes client dependencies, otherwise the tests would fail due to missing classes.

Comment (aloubyansky) [reply test-framework/kubernetes-client/pom.xml:27]: Excluding that dependency results in failures like
````
Caused by: java.lang.NoClassDefFoundError: okio/Sink
	at io.fabric8.kubernetes.client.server.mock.KubernetesServer.before(KubernetesServer.java:69)
	at io.quarkus.test.kubernetes.client.KubernetesServerTestResource.initServer(KubernetesServerTestResource.java:39)
	at io.quarkus.test.kubernetes.client.AbstractKubernetesTestResource.start(AbstractKubernetesTestResource.java:25)
	at io.quarkus.test.common.TestResourceManager.lambda$start$1(TestResourceManager.java:105)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassNotFoundException: okio.Sink
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
````

**Review by gsmet — APPROVED** [review]
